File: dsa-05-main/cpp/day06/singly_list.cpp
#include <iostream>
using namespace std;

class singly_list; // forward declaration class

class node {
private:
	int data;
	node* next;
public:
	node() {
		data = 0;
		next = NULL;
	}
	node(int val) {
		data = val;
		next = NULL;
	}
	friend class singly_list;
};

class singly_list {
private:
	node* head;
public:
	singly_list() {
		head = NULL;
	}
	~singly_list() {
		del_all();
	}
	void add_first(int val) {
		node* newnode = new node(val);
		newnode->next = head;
		head = newnode;
	}
	void add_last(int val) {
		node* newnode, * trav;
		// alloc new node and init it
		newnode = new node(val);
		// special case: if list is empty, newnode itself is first node
		if (head == NULL)
			head = newnode;
		else {
			// traverse till last node
			trav = head;
			while (trav->next != NULL)
				trav = trav->next;
			// add new node to the next of last node
			trav->next = newnode;
		}
	}
	void add_at_pos(int pos, int val) {
		node* newnode, * trav;
		// special 1 & 2: if list is empty, new node is first node | user give pos = 1 or less than that, new node should be first
		if (head == NULL || pos <= 1) {
			add_first(val);
			return;
		}
		// alloc node and init it
		newnode = new node(val);
		// trav till pos-1
		trav = head;
		for (int i = 1; i < pos - 1; i++) {
			// special3: if pos is beyond size of list, add node at the end.
			if (trav->next == NULL)
				break;
			trav = trav->next;
		}
		// newnode'next to trav's next
		newnode->next = trav->next;
		// previous (trav) 'next to newnode
		trav->next = newnode;
	}
	void display() {
		cout << "LIST: ";
		node* trav = head;
		while (trav != NULL) {
			cout << trav->data << ", ";
			trav = trav->next;
		}
		cout << endl;
	}
	void del_first() {
		node* temp;
		if (head != NULL) {
			temp = head;
			head = head->next;
			delete temp;
		}
	}
	void del_all() {
		while (head != NULL)
			del_first();
	}
	void del_at_pos(int pos) {
		node* temp, *trav;
		int i;
		//special 1 & 2: if list is empty or pos == 1, delete the first node
		if (head == NULL || pos == 1)
			del_first();
		else {
			//1. trav till pos-1
			trav = head;
			for (i = 1; i < pos - 1; i++) {
				//special 3: if pos is beyond size of list, don't do anything.
				if (trav->next == NULL)
					return;
				trav = trav->next;
			}
			//2. get pointer temp to the next node (to be deleted)
			temp = trav->next;
			//3. trav's next to temp's next
			trav->next = temp->next;
			//4. delete temp node
			delete temp;
		}
	}
};

int main() {
	singly_list l1;
	l1.add_last(10);
	l1.add_last(20);
	l1.add_last(30);
	l1.add_first(40);
	l1.add_at_pos(3, 50);
	l1.add_at_pos(1, 60);
	l1.add_at_pos(8, 70);
	l1.display();
	l1.del_first();
	l1.display();
	l1.del_first();
	l1.display();
	l1.del_at_pos(3);
	l1.display();
	l1.del_at_pos(4);
	l1.display();
	l1.del_at_pos(5);
	l1.display();
	//l1.del_all();
	//l1.display();
	return 0;
}

File: dsa-05-main/python/day06/singly_list.py
class Node:
	def __init__(self, val=0):
		self.data = val
		self.next = None


class SinglyList:

	def __init__(self):
		self.head = None

	# time complexity - O(1)
	def addFirst(self, val):
		newNode = Node(val)
		newNode.next = self.head
		self.head = newNode


	# time complexity - O(n)
	def addLast(self, val):
		newNode = Node(val)
		# special case: if list is empty
		if self.head is None:
			self.head = newNode # newNode itself is first node of list
		else: # general case: if list already have few nodes
			# traverse till last node
			trav = self.head
			while trav.next is not None:
				trav = trav.next
			# add address of newNode into last node's next
			trav.next = newNode


	# time complexity - O(n)
	def display(self):
		print("List: ", end='')
		trav = self.head
		while trav is not None:
			print(trav.data, end=", ")
			trav = trav.next
		print()

	# time complexity - O(n)
	def addAtPos(self, val, pos):
		# spl 4. invalid pos( < 1)
		if pos < 1:
			raise Exception("Invalid position " + str(pos))
		# spl 1. if list is empty OR spl 2. add at pos = 1
		if self.head is None or pos == 1:
			self.addFirst(val)
		else:
			# create a new node
			newNode = Node(val)
			# traverse till pos - 1
			trav = self.head
			for i in range(1, pos-1):
				# spl 3. if pos to add is beyond length of list, throw error
				trav = trav.next
				if trav is None:
					raise Exception("Invalid position " + str(pos))
			# add node at that pos
			newNode.next = trav.next
			trav.next = newNode


	# time complexity - O(1)
	def delFirst(self):
		if self.head is None:
			raise Exception("List is empty.")
		self.head = self.head.next

	def delAtPos(self, pos):
		# spl 3: if pos < 1, throw exception
		if pos < 1:
			raise Exception("Invalid position " + str(pos))
		# spl 1: list is empty & 2: pos == 1
		if self.head is None or pos == 1:
			self.delFirst()
		else:
			trav = self.head
			for i in range(1, pos-1):
				trav = trav.next
				# spl 4: if pos > length, throw exception
				if trav is None:
					raise Exception("Invalid position " + str(pos))
			temp = trav.next
			# spl 5: pos == length, if temp is null, throw exception
			if temp is None:
				raise Exception("Invalid position " + str(pos))
			trav.next = temp.next

	# time complexity - O(n)
	def delAll(self):
		while self.head is not None:
			self.delFirst()
		# head = null;

	# time complexity - O(n)
	def delLast(self):
		# spl1: list is empty, spl2: list have only one node.
		if self.head is None or self.head.next is None:
			self.delFirst()
		else:
			temp = None
			trav = self.head
			while trav.next is not None:
				temp = trav
				trav = trav.next
			temp.next = None


def main():
	l1 = SinglyList()
	l1.addLast(11)
	l1.addLast(22)
	l1.addLast(33)
	l1.addFirst(44)
	l1.display()
	l1.addAtPos(55, 2)
	l1.display()
	l1.delFirst()
	l1.display()
	l1.delLast()
	l1.display()
	l1.delAtPos(3)
	l1.display()



if __name__ == "__main__":
	main()

